### PR TITLE
CVSL-199 expand additional conditions data into placeholders

### PR DESCRIPTION
**Pull request by timharrison-moj**
This PR does a few things:
* Adds templates into the configuration for additional conditions which require input.
* Expands placeholders with their values prior to printing a licence
* Caters for data item case (lower, upper, capitalise) in expanded values
* Caters for lists of data items, formatted with commas and/or 'and' between values.
* Tests for various types of additional conditions and their value replacement.

### CODE REVIEW

File: server/utils/conditionsProvider.ts [hunk 1]
```diff
@@ -35,3 +37,156 @@ export function getGroupedAdditionalConditions(): Record<string, unknown>[] {
...
+  const expandedConditions: string[] = []
+
+  conditions?.forEach(condition => {
+    const configCondition = getAdditionalConditionByCode(condition.code)
+    if (configCondition?.requiresInput) {
+      const placeholders = getPlaceholderNames(configCondition.tpl as string)
+      const inputConfig = configCondition.inputs as unknown as Record<string, unknown>[]
+      let thisConditionText = configCondition.tpl as string
+
+      for (let i = 0; i < placeholders.length; i += 1) {
```
Comment (smcveigh941) [inline server/utils/conditionsProvider.ts:56]: `placeholders.forEach()` ?

File: server/utils/conditionsProvider.ts [hunk 2]
```diff
@@ -35,3 +37,156 @@ export function getGroupedAdditionalConditions(): Record<string, unknown>[] {
...
+          // Single matching value
+          const rules = inputConfig.find(item => item.name === ph)
+          const { value } = matchingDataItems[0]
+          thisConditionText = replacePlaceholderWithValue(
+            ph,
+            thisConditionText,
+            adjustCase(rules.case as string, value)
+          )
+        } else {
+          // List of values for this placeholder
```
Comment (smcveigh941) [inline server/utils/conditionsProvider.ts:76]: I might be inclined to group fields by the same name into one single string before doing this processing. Doing that preprocessing would mean you don't need your else block here. I have this function in my PR to group fields into one string. But not to worry

```
    const map = new Map()
    data.forEach(item => {
      const collection = map.get(item.field)
      if (!collection) {
        map.set(item.field, [item])
      } else {
        collection.push(item)
      }
    })
    return Array.from(map, ([fieldName, items]) => ({ fieldName, items })).map(groupedDataItems => {
      return groupedDataItems.items.map((dataItem: AdditionalConditionData) => dataItem.value).join(', ')
    })
```

File: server/utils/conditionsProvider.ts [hunk 3]
```diff
@@ -35,3 +37,156 @@ export function getGroupedAdditionalConditions(): Record<string, unknown>[] {
...
+}
+
+/**
+ * Formats a list of data items according to rules (add commas, and)
+ * @param matchingDataItems
+ */
+export const produceValueAsFormattedList = (matchingDataItems: AdditionalConditionData[]): string => {
+  let value = ''
+  let valueCounter = 1
+  for (let i = 0; i < matchingDataItems.length; i += 1) {
```
Comment (smcveigh941) [inline server/utils/conditionsProvider.ts:105]: matchingDataItems.forEach(item => {})

File: server/utils/conditionsProviderExpand.test.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+          code: 'ed607a91-fe3a-4816-8eb9-b447c945935c',
+          category: 'Possession, ownership, control or inspection of specified items or documents',
+          sequence: 1,
+          text: 'Not to own or use a camera without the prior approval of your supervising officer.',
+          data: [],
+        },
+      ]
+      const listOfConditions = expandAdditionalConditions(conditions)
+      expect(listOfConditions).toHaveLength(1)
+      expect(listOfConditions[0]).toEqual(conditions[0].text)
```
Comment (smcveigh941) [inline server/utils/conditionsProviderExpand.test.ts:25]: Always inclined to use string literals over variables in assertions

File: server/utils/conditionsProviderExpand.test.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        },
+      ]
+      const listOfConditions = expandAdditionalConditions(conditions)
+      expect(listOfConditions).toHaveLength(1)
+      expect(listOfConditions[0]).toEqual(
+        'To comply with any requirements specified by your supervising officer for the purpose of ensuring that you address your alcohol, drug, sexual, violent, gambling and anger problems.'
+      )
+    })
+  })
+})
```
Comment (smcveigh941) [inline server/utils/conditionsProviderExpand.test.ts:151]: Good tests!

**Review by smcveigh941 — APPROVED** [review]
Looks sensible to me, some suggestions

File: server/utils/conditionsProvider.ts (hunk 1 above)
Comment (timharrison-moj) [reply server/utils/conditionsProvider.ts:56]: I did this first time around, however it never goes into the loop! Something odd going on with Array.forEach in these circumstances, which I was thinking I would come back to once the code and tests were in and working.

File: server/utils/conditionsProvider.ts (hunk 3 above)
Comment (timharrison-moj) [reply server/utils/conditionsProvider.ts:105]: Its the same thing here - I had it like this originally, but it never went into the forEach loop, despite it having values.

File: server/utils/conditionsProviderExpand.test.ts (hunk 4 above)
Comment (timharrison-moj) [reply server/utils/conditionsProviderExpand.test.ts:25]: In this case, its taking the .text verbatim - so should be identical. Agree where there are changes being made.

File: server/utils/conditionsProviderExpand.test.ts (hunk 5 above)
Comment (timharrison-moj) [reply server/utils/conditionsProviderExpand.test.ts:151]: Thanks!

File: server/utils/conditionsProvider.ts (hunk 1 above)
Comment (timharrison-moj) [reply server/utils/conditionsProvider.ts:56]: Any ideas appreciated why the .forEach might not work here?

File: server/utils/conditionsProvider.ts (hunk 2 above)
Comment (timharrison-moj) [reply server/utils/conditionsProvider.ts:76]: I'll take a another look at this now.

File: server/utils/conditionsProvider.ts (hunk 1 above)
Comment (timharrison-moj) [reply server/utils/conditionsProvider.ts:56]: Done now - next PR.

**Pull request merged**